packages/salesforcedx-vscode-soql/test/vscode-integration/editor/soqlEditorInstance.test.ts

```diff
@@ -0,0 +1,157 @@
+      .withArgs('defaultusername')
+      .returns(testData.username);
+    docProviderDisposable = vscode.workspace.registerTextDocumentContentProvider(
+      'sfdc-test',
+      new MockTextDocumentProvider()
+    );
+    mockTextDocument = await vscode.workspace.openTextDocument(
+      vscode.Uri.parse('sfdc-test:mocksoql.soql')
+    );
+    mockWebviewPanel = vscode.window.createWebviewPanel(
+      'mockWebviewPanel',
+      'Mock Webview Panel',
+      vscode.ViewColumn.One,
+      { enableScripts: true }
+    );
+    mockWebviewPanel.webview.html = createMessagingWebviewContent();
+    instance = new TestSoqlEditorInstance(
+      mockTextDocument,
+      mockWebviewPanel,
+      new vscode.CancellationTokenSource().token
+    );
+  });
+  afterEach(() => {
+    mockWebviewPanel.dispose();
+    docProviderDisposable.dispose();
+    $$.SANDBOX.restore();
+    sandbox.restore();
+  });
+
+  it('responds to sobjects_request with a list of sobjects', async () => {
+    sandbox
+      .stub(OrgAuthInfo, 'getDefaultUsernameOrAlias')
+      .returns(testData.username);
+    sandbox.stub(OrgAuthInfo, 'getConnection').returns(mockConnection);
+    const sobjectNames = ['A', 'B'];
+    sandbox
+      .stub(SObjectService.prototype, 'retrieveSObjectNames')
+      .resolves(sobjectNames);
+
+    const expectedMessage = {
+      type: 'sobjects_response',
+      message: sobjectNames
+    };
+    const postMessageSpy = sandbox.spy(mockWebviewPanel.webview, 'postMessage');
+
+    instance.sendEvent({ type: 'sobjects_request' });
+    // above function has nested async message passing; wait a bit
+    await waitForAsync(50);
+
+    assert(postMessageSpy.calledWith(expectedMessage));
+  });
+
+  it('responds to sobject_metadata_request with SObject metadata', async () => {
+    sandbox
+      .stub(OrgAuthInfo, 'getDefaultUsernameOrAlias')
+      .returns(testData.username);
+    sandbox.stub(OrgAuthInfo, 'getConnection').returns(mockConnection);
+    const fakeSObject = { name: 'A' };
+    sandbox
+      .stub(SObjectService.prototype, 'describeSObject')
+      .resolves(fakeSObject);
+
+    const expectedMessage = {
+      type: 'sobject_metadata_response',
+      message: fakeSObject
+    };
+    const postMessageSpy = sandbox.spy(mockWebviewPanel.webview, 'postMessage');
+
+    instance.sendEvent({ type: 'sobject_metadata_request' });
+    // above function has nested async message passing; wait a bit
+    await waitForAsync(50);
+
+    assert(postMessageSpy.calledWith(expectedMessage));
+  });
+});
+
+class MockTextDocumentProvider implements vscode.TextDocumentContentProvider {
+  public provideTextDocumentContent(
+    uri: vscode.Uri,
+    token: vscode.CancellationToken
+  ): string {
+    return 'SELECT A FROM B';
+  }
+}
+
+class TestSoqlEditorInstance extends SOQLEditorInstance {
+  public sendEvent(event: any) {
+    this.onDidRecieveMessageHandler({
+      type: event.type,
+      message: event.message
+    });
+  }
+}
```

Comment:
This subclass allows us to simulate message passing from the UI with minimal setup.